component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/voiceotp/VoiceOTPAuthenticator.java

```diff
@@ -0,0 +1,2034 @@
+            }
+            FederatedAuthenticatorUtil.setUsernameFromFirstStep(context);
+            username = String.valueOf(context.getProperty(VoiceOTPConstants.USER_NAME));
+            authenticatedUser = (AuthenticatedUser) context.getProperty(VoiceOTPConstants.AUTHENTICATED_USER);
+            // Find the authenticated user.
+            if (authenticatedUser == null) {
+                LOG.error("Authentication failed: Could not find the authenticated user.");
+                throw new AuthenticationFailedException
+                        ("Authentication failed: Cannot proceed further without identifying the user.");
+            }
+            boolean isVoiceOTPMandatory = VoiceOTPUtils.isVoiceOTPMandatory(context);
+            boolean isUserExists = FederatedAuthenticatorUtil.isUserExistInUserStore(username);
+            String queryParams = FrameworkUtils.getQueryStringWithFrameworkContextId(context.getQueryParams(),
+                    context.getCallerSessionKey(), context.getContextIdentifier());
+            // This multi option URI is used to navigate back to multi option page to select a different
+            // authentication option from Voice OTP pages.
+            String multiOptionURI = getMultiOptionURIQueryParam(request);
+            if (StringUtils.isNotEmpty(multiOptionURI)) {
+                queryParams += multiOptionURI;
+            }
+            String errorPage = getErrorPage(context);
+            // Voice OTP authentication is mandatory and user doesn't disable Voice OTP claim in user's profile.
+            if (isVoiceOTPMandatory) {
+
+                LOG.debug("Voice OTP is mandatory. Hence processing in mandatory path.");
+
+                processVoiceOTPMandatoryCase(context, request, response, queryParams, username, isUserExists);
+            // Checks whether local user exists and the voice otp is not disabled for the user.
+            } else if (isUserExists && !VoiceOTPUtils.isVoiceOTPDisableForLocalUser(username, context)) {
+                // Trigger retry flow when it is not request for resending OTP, it is not mobile number update failure
+                // or the user is a local user and the user account is locked.
+                if ((context.isRetrying() && !Boolean.parseBoolean(request.getParameter(VoiceOTPConstants.RESEND))
+                        && !isMobileNumberUpdateFailed(context)) || (VoiceOTPUtils.isLocalUser(context) &&
+                        VoiceOTPUtils.isAccountLocked(authenticatedUser))) {
+
+                    LOG.debug("Triggering Voice OTP retry flow.");
+
+                    checkStatusCode(response, context, queryParams, errorPage);
+                } else {
+                    mobileNumber = getMobileNumber(request, response, context, username, queryParams);
+                    if (StringUtils.isNotEmpty(mobileNumber)) {
+                        proceedWithOTP(response, context, errorPage, mobileNumber, queryParams, username);
```

Comment:
Is it possible for `mobileNumber ` to be empty here ? If so where do we handle this ?

Reply:
Yes. It is handled inside the getMobileNumber method since it is calling redirectToMobileNoReqPage function if the mobile number is empty.